File: frontend/src/types/itsm.ts
/**
 * TypeScript types + Zod schemas untuk ITSM modules
 * State machine transitions di-mirror dari backend untuk UX feedback
 */
import { z } from 'zod'

// ── Status enums ──────────────────────────────────────────────
export const INCIDENT_STATUSES  = ['Open', 'In Progress', 'Resolved', 'Closed', 'Cancelled'] as const
export const INCIDENT_PRIORITIES = ['Low', 'Medium', 'High', 'Critical'] as const
export const INCIDENT_SEVERITIES = ['S1', 'S2', 'S3', 'S4'] as const

export const CHANGE_STATUSES    = ['Draft', 'Submitted', 'Approved', 'Rejected', 'In Progress', 'Completed', 'Cancelled'] as const
export const CHANGE_PRIORITIES  = ['Low', 'Medium', 'High', 'Critical'] as const
export const CHANGE_TYPES       = ['Standard', 'Normal', 'Emergency'] as const

export const PROBLEM_STATUSES   = ['Open', 'Under Investigation', 'Known Error', 'Resolved', 'Closed'] as const
export const REQUEST_STATUSES   = ['Draft', 'Submitted', 'Approved', 'Rejected', 'In Progress', 'Completed', 'Cancelled'] as const
export const REQUEST_PRIORITIES = ['Low', 'Medium', 'High'] as const
export const REQUEST_TYPES      = ['New Asset', 'Repair', 'Replacement', 'Software', 'Access', 'Other'] as const

// ── State machine (mirror dari backend) ──────────────────────
export const INCIDENT_TRANSITIONS: Record<string, string[]> = {
  'Open':        ['In Progress', 'Cancelled'],
  'In Progress': ['Resolved', 'Cancelled'],
  'Resolved':    ['Closed', 'Open'],
  'Closed':      [],
  'Cancelled':   [],
}

export const CHANGE_TRANSITIONS: Record<string, string[]> = {
  'Draft':       ['Submitted', 'Cancelled'],
  'Submitted':   ['Approved', 'Rejected', 'Cancelled'],
  'Approved':    ['In Progress', 'Cancelled'],
  'Rejected':    ['Draft'],
  'In Progress': ['Completed', 'Cancelled'],
  'Completed':   [],
  'Cancelled':   [],
}

export const PROBLEM_TRANSITIONS: Record<string, string[]> = {
  'Open':                ['Under Investigation', 'Closed'],
  'Under Investigation': ['Known Error', 'Resolved', 'Closed'],
  'Known Error':         ['Resolved', 'Closed'],
  'Resolved':            ['Closed', 'Open'],
  'Closed':              [],
}

export const REQUEST_TRANSITIONS: Record<string, string[]> = {
  'Draft':       ['Submitted', 'Cancelled'],
  'Submitted':   ['Approved', 'Rejected', 'Cancelled'],
  'Approved':    ['In Progress', 'Cancelled'],
  'Rejected':    ['Draft'],
  'In Progress': ['Completed', 'Cancelled'],
  'Completed':   [],
  'Cancelled':   [],
}

// ── Zod schemas ───────────────────────────────────────────────
export const IncidentSchema = z.object({
  title:           z.string().min(1, 'Title is required').max(200),
  description:     z.string().max(5000).optional(),
  priority:        z.enum(INCIDENT_PRIORITIES).default('Medium'),
  severity:        z.enum(INCIDENT_SEVERITIES).default('S3'),
  status:          z.enum(INCIDENT_STATUSES).default('Open'),
  asset_id:        z.number().int().positive().nullable().optional(),
  assigned_to:     z.number().int().positive().nullable().optional(),
  resolution_note: z.string().max(5000).optional(),
})

export const ChangeSchema = z.object({
  title:         z.string().min(1, 'Title is required').max(200),
  description:   z.string().max(5000).optional(),
  change_type:   z.enum(CHANGE_TYPES).default('Normal'),
  priority:      z.enum(CHANGE_PRIORITIES).default('Medium'),
  status:        z.enum(CHANGE_STATUSES).default('Draft'),
  asset_id:      z.number().int().positive().nullable().optional(),
  planned_start: z.string().optional(),
  planned_end:   z.string().optional(),
  rollback_plan: z.string().max(5000).optional(),
  notes:         z.string().max(5000).optional(),
})

export const ProblemSchema = z.object({
  title:       z.string().min(1, 'Title is required').max(200),
  description: z.string().max(5000).optional(),
  root_cause:  z.string().max(5000).optional(),
  workaround:  z.string().max(5000).optional(),
  status:      z.enum(PROBLEM_STATUSES).default('Open'),
  asset_id:    z.number().int().positive().nullable().optional(),
  assigned_to: z.number().int().positive().nullable().optional(),
})

export const RequestSchema = z.object({
  title:        z.string().min(1, 'Title is required').max(200),
  description:  z.string().max(5000).optional(),
  request_type: z.enum(REQUEST_TYPES).default('Other'),
  priority:     z.enum(REQUEST_PRIORITIES).default('Medium'),
  status:       z.enum(REQUEST_STATUSES).default('Draft'),
  asset_id:     z.number().int().positive().nullable().optional(),
  notes:        z.string().max(5000).optional(),
})

export type IncidentForm = z.infer<typeof IncidentSchema>
export type ChangeForm   = z.infer<typeof ChangeSchema>
export type ProblemForm  = z.infer<typeof ProblemSchema>
export type RequestForm  = z.infer<typeof RequestSchema>

// ── Badge color maps ──────────────────────────────────────────
export const STATUS_BADGE: Record<string, 'success' | 'info' | 'warning' | 'error' | 'neutral'> = {
  // Incident
  'Open':                  'error',
  'In Progress':           'warning',
  'Resolved':              'info',
  'Closed':                'success',
  'Cancelled':             'neutral',
  // Change/Request
  'Draft':                 'neutral',
  'Submitted':             'info',
  'Approved':              'success',
  'Rejected':              'error',
  'Completed':             'success',
  // Problem
  'Under Investigation':   'warning',
  'Known Error':           'error',
}

export const PRIORITY_BADGE: Record<string, 'success' | 'info' | 'warning' | 'error' | 'neutral'> = {
  'Low':      'neutral',
  'Medium':   'info',
  'High':     'warning',
  'Critical': 'error',
}

File: backend/app/models/itsm.py
"""Model ITSM: Change, Incident, Problem, Request."""
import enum
from datetime import datetime, timezone
from sqlalchemy import String, Integer, DateTime, Text, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from ..extensions import db


# ── Enums ─────────────────────────────────────────────────────

class ChangeStatus(str, enum.Enum):
    DRAFT       = "Draft"
    SUBMITTED   = "Submitted"
    APPROVED    = "Approved"
    REJECTED    = "Rejected"
    IN_PROGRESS = "In Progress"
    COMPLETED   = "Completed"
    CANCELLED   = "Cancelled"

class ChangePriority(str, enum.Enum):
    LOW      = "Low"
    MEDIUM   = "Medium"
    HIGH     = "High"
    CRITICAL = "Critical"

class ChangeType(str, enum.Enum):
    STANDARD  = "Standard"
    NORMAL    = "Normal"
    EMERGENCY = "Emergency"

class IncidentStatus(str, enum.Enum):
    OPEN        = "Open"
    IN_PROGRESS = "In Progress"
    RESOLVED    = "Resolved"
    CLOSED      = "Closed"
    CANCELLED   = "Cancelled"

class IncidentPriority(str, enum.Enum):
    LOW      = "Low"
    MEDIUM   = "Medium"
    HIGH     = "High"
    CRITICAL = "Critical"

class IncidentSeverity(str, enum.Enum):
    S1 = "S1"
    S2 = "S2"
    S3 = "S3"
    S4 = "S4"

class ProblemStatus(str, enum.Enum):
    OPEN              = "Open"
    UNDER_INVESTIGATION = "Under Investigation"
    KNOWN_ERROR       = "Known Error"
    RESOLVED          = "Resolved"
    CLOSED            = "Closed"

class RequestStatus(str, enum.Enum):
    DRAFT       = "Draft"
    SUBMITTED   = "Submitted"
    APPROVED    = "Approved"
    REJECTED    = "Rejected"
    IN_PROGRESS = "In Progress"
    COMPLETED   = "Completed"
    CANCELLED   = "Cancelled"

class RequestPriority(str, enum.Enum):
    LOW    = "Low"
    MEDIUM = "Medium"
    HIGH   = "High"

class RequestType(str, enum.Enum):
    NEW_ASSET   = "New Asset"
    REPAIR      = "Repair"
    REPLACEMENT = "Replacement"
    SOFTWARE    = "Software"
    ACCESS      = "Access"
    OTHER       = "Other"


# ── Change ────────────────────────────────────────────────────

class Change(db.Model):
    __tablename__ = "changes"
    id:             Mapped[int]  = mapped_column(Integer, primary_key=True)
    change_number:  Mapped[str]  = mapped_column(String(20), unique=True, nullable=False)
    title:          Mapped[str]  = mapped_column(String(200), nullable=False)
    description:    Mapped[str | None] = mapped_column(Text)
    change_type:    Mapped[ChangeType]     = mapped_column(Enum(ChangeType,     name="change_type"),     default=ChangeType.NORMAL)
    priority:       Mapped[ChangePriority] = mapped_column(Enum(ChangePriority, name="change_priority"), default=ChangePriority.MEDIUM)
    status:         Mapped[ChangeStatus]   = mapped_column(Enum(ChangeStatus,   name="change_status"),   default=ChangeStatus.DRAFT)
    asset_id:       Mapped[int | None] = mapped_column(Integer, db.ForeignKey("assets.id"))
    requested_by:   Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    approved_by:    Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    planned_start:  Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    planned_end:    Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    actual_start:   Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    actual_end:     Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    rollback_plan:  Mapped[str | None] = mapped_column(Text)
    notes:          Mapped[str | None] = mapped_column(Text)
    created_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    deleted_at:     Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


# ── Incident ──────────────────────────────────────────────────

class Incident(db.Model):
    __tablename__ = "incidents"
    id:              Mapped[int]  = mapped_column(Integer, primary_key=True)
    incident_number: Mapped[str]  = mapped_column(String(20), unique=True, nullable=False)
    title:           Mapped[str]  = mapped_column(String(200), nullable=False)
    description:     Mapped[str | None] = mapped_column(Text)
    priority:        Mapped[IncidentPriority] = mapped_column(Enum(IncidentPriority, name="incident_priority"), default=IncidentPriority.MEDIUM)
    severity:        Mapped[IncidentSeverity] = mapped_column(Enum(IncidentSeverity, name="incident_severity"), default=IncidentSeverity.S3)
    status:          Mapped[IncidentStatus]   = mapped_column(Enum(IncidentStatus,   name="incident_status"),   default=IncidentStatus.OPEN)
    asset_id:        Mapped[int | None] = mapped_column(Integer, db.ForeignKey("assets.id"))
    reported_by:     Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    assigned_to:     Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    resolved_by:     Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    reported_at:     Mapped[datetime]   = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    resolved_at:     Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    closed_at:       Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    resolution_note: Mapped[str | None] = mapped_column(Text)
    created_at:      Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at:      Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    deleted_at:      Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


# ── Problem ───────────────────────────────────────────────────

class Problem(db.Model):
    __tablename__ = "problems"
    id:             Mapped[int]  = mapped_column(Integer, primary_key=True)
    problem_number: Mapped[str]  = mapped_column(String(20), unique=True, nullable=False)
    title:          Mapped[str]  = mapped_column(String(200), nullable=False)
    description:    Mapped[str | None] = mapped_column(Text)
    root_cause:     Mapped[str | None] = mapped_column(Text)
    workaround:     Mapped[str | None] = mapped_column(Text)
    status:         Mapped[ProblemStatus] = mapped_column(Enum(ProblemStatus, name="problem_status"), default=ProblemStatus.OPEN)
    asset_id:       Mapped[int | None] = mapped_column(Integer, db.ForeignKey("assets.id"))
    reported_by:    Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    assigned_to:    Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    resolved_by:    Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    resolved_at:    Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    deleted_at:     Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class IncidentProblem(db.Model):
    __tablename__ = "incident_problems"
    incident_id: Mapped[int] = mapped_column(Integer, db.ForeignKey("incidents.id", ondelete="CASCADE"), primary_key=True)
    problem_id:  Mapped[int] = mapped_column(Integer, db.ForeignKey("problems.id",  ondelete="CASCADE"), primary_key=True)


# ── Request ───────────────────────────────────────────────────

class Request(db.Model):
    __tablename__ = "requests"
    id:             Mapped[int]  = mapped_column(Integer, primary_key=True)
    request_number: Mapped[str]  = mapped_column(String(20), unique=True, nullable=False)
    title:          Mapped[str]  = mapped_column(String(200), nullable=False)
    description:    Mapped[str | None] = mapped_column(Text)
    request_type:   Mapped[RequestType]     = mapped_column(Enum(RequestType,     name="request_type"),     default=RequestType.OTHER)
    priority:       Mapped[RequestPriority] = mapped_column(Enum(RequestPriority, name="request_priority"), default=RequestPriority.MEDIUM)
    status:         Mapped[RequestStatus]   = mapped_column(Enum(RequestStatus,   name="request_status"),   default=RequestStatus.DRAFT)
    asset_id:       Mapped[int | None] = mapped_column(Integer, db.ForeignKey("assets.id"))
    requested_by:   Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    approved_by:    Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    fulfilled_by:   Mapped[int | None] = mapped_column(Integer, db.ForeignKey("accounts.id"))
    approved_at:    Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    fulfilled_at:   Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    notes:          Mapped[str | None] = mapped_column(Text)
    created_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at:     Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    deleted_at:     Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
